types/int.go

```diff
@@ -0,0 +1,452 @@
+	return integer{*big.NewInt(n)}
+}
+
+func (i integer) bigInt() *big.Int {
+	return new(big.Int).Set(&(i.i))
+}
+
+func newIntegerFromBigInt(i *big.Int) integer {
+	return integer{*i}
+}
+
+func newIntegerFromString(s string) (res integer, ok bool) {
+	i, ok := new(big.Int).SetString(s, 0)
+	if !ok {
+		return
+	}
+	return integer{*i}, ok
+}
+
+func newIntegerWithDecimal(n int64, dec int) (res integer) {
+	if dec < 0 {
+		return
+	}
+	exp := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(dec)), nil)
+	i := new(big.Int)
+	return integer{*(i.Mul(big.NewInt(n), exp))}
+}
+
+func zeroInt() integer { return integer{*big.NewInt(0)} }
+func oneInt() integer  { return integer{*big.NewInt(1)} }
+
+func (i integer) toInt64() int64 {
+	if !(&(i.i)).IsInt64() {
+		panic("Out of bound in Int64()")
+	}
+	return (&(i.i)).Int64()
+}
+
+func (i integer) toUint64() uint64 {
+	if !(&(i.i)).IsUint64() {
+		panic("Out of bound in Int64()")
+	}
+	return (&(i.i)).Uint64()
+}
+
+func (i integer) isZero() bool { return (&(i.i)).Sign() == 0 }
+
+func (i integer) sign() int { return (&(i.i)).Sign() }
+
+func (i integer) equal(i2 integer) bool { return (&(i.i)).Cmp(&(i2.i)) == 0 }
+
+func (i integer) gt(i2 integer) bool { return (&(i.i)).Cmp(&(i2.i)) == 1 }
+
+func (i integer) lt(i2 integer) bool { return (&(i.i)).Cmp(&(i2.i)) == -1 }
+
+func (i integer) add(i2 integer) integer { return integer{*new(big.Int).Add(&(i.i), &(i2.i))} }
+
+func (i integer) sub(i2 integer) integer { return integer{*new(big.Int).Sub(&(i.i), &(i2.i))} }
+
+func (i integer) mul(i2 integer) integer { return integer{*new(big.Int).Mul(&(i.i), &(i2.i))} }
+
+func (i integer) div(i2 integer) integer { return integer{*new(big.Int).Div(&(i.i), &(i2.i))} }
+
+func (i integer) mod(i2 integer) integer { return integer{*new(big.Int).Mod(&(i.i), &(i2.i))} }
+
+func (i integer) neg() integer { return integer{*new(big.Int).Neg((&(i.i)))} }
+
+func (i integer) String() string { return (&(i.i)).String() }
+
+func (i integer) bitLen() int { return (&(i.i)).BitLen() }
+
+// MarshalAmino for custom encoding scheme
+func (i integer) MarshalAmino() (string, error) {
+	bz, err := (&(i.i)).MarshalText()
+	return string(bz), err
+}
+
+// UnmarshalAmino for custom decoding scheme
+func (i *integer) UnmarshalAmino(text string) (err error) {
+	tempInt := new(big.Int)
+	err = tempInt.UnmarshalText([]byte(text))
+	if err != nil {
+		return
+	}
+	i.i = *tempInt
+	return nil
+}
+
+// MarshalJSON for custom encodig scheme
+func (i integer) MarshalJSON() ([]byte, error) {
+	return (&(i.i)).MarshalText()
+}
+
+// UnmarshalJSON for custom decoding scheme
+func (i *integer) UnmarshalJSON(bz []byte) (err error) {
+	tempInt := new(big.Int)
+	err = tempInt.UnmarshalText([]byte(bz))
+	if err != nil {
+		return
+	}
+	i.i = *tempInt
+	return nil
+}
+
+// Int wraps integer with 256 bit range bound
+// Checks overflow, underflow and division by zero
+// Exists in range from -(2^255-1) to 2^255-1
+type Int struct {
+	i integer
```

Comment:
I don't entirely understand why we need a double wrapper - can't we just put `big.Int` as an unexported field in the `sdk.Int` type?

Reply:
I guess it's cleaner with both `sdk.Int` and `sdk.Uint`?